File: src/VBNetFx/FacessoData/SQLServer/SPAccess_WorkGroups.vb
Imports Activedev
Imports Facesso
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.Common

Partial Public NotInheritable Class SPAccess

    Public Function WorkGroups_DoesWorkGroupNumberExist(ByVal IDSubsidiary As Integer, _
                                                        ByVal WorkGroupNumber As Integer, _
                                                        ByVal ExcludeIDWorkGroup As ADDBNullable(Of Integer)) _
                                                        As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_DoesWorkGroupNumberExist", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@WorkGroupNumber", SqlDbType.Int).Value = WorkGroupNumber
                .Add("@ExcludeIDWorkGroup", SqlDbType.Int).Value = ExcludeIDWorkGroup.Value
                .Add("@DoesExist", SqlDbType.Bit)
                .Item("@DoesExist").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CBool(locCmd.Parameters("@DoesExist").Value)
        End Using
    End Function

    Public Function WorkGroups_DoesWorkGroupNameExist(ByVal IDSubsidiary As Integer, _
                                                    ByVal WorkGroupName As String, _
                                                    ByVal ExcludeIDWorkGroup As ADDBNullable(Of Integer)) _
                                                    As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_DoesWorkGroupNameExist", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@WorkGroupName", SqlDbType.NVarChar, 20).Value = WorkGroupName
                .Add("@ExcludeIDWorkGroup", SqlDbType.Int).Value = ExcludeIDWorkGroup.Value
                .Add("@DoesExist", SqlDbType.Bit)
                .Item("@DoesExist").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CBool(locCmd.Parameters("@DoesExist").Value)
        End Using
    End Function

    Public Function WorkGroups_Add(ByVal wgi As WorkGroupInfo, ByVal CreatedByIDUser As Integer) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_Add", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = wgi.IDCostCenter
                .Add("@WorkGroupNumber", SqlDbType.Int).Value = wgi.WorkGroupNumber
                .Add("@WorkGroupName", SqlDbType.NVarChar, 100).Value = wgi.WorkGroupName
                .Add("@WorkGroupDescription", SqlDbType.NVarChar, 4000).Value = wgi.WorkGroupDescription.Value
                .Add("@IsActive", SqlDbType.Bit).Value = wgi.IsActive
                .Add("@IsPeaceWork", SqlDbType.Bit).Value = wgi.IsPeaceWork
                .Add("@IsConceptional", SqlDbType.Bit).Value = wgi.IsConceptional
                .Add("@OrdinalNo", SqlDbType.Int).Value = wgi.OrdinalNo
                .Add("@TimeSettingDetails", SqlDbType.Xml).Value = wgi.TimeSettingDetails.XMLString
                .Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate
                .Add("@CreatedByIDUser", SqlDbType.Int).Value = CreatedByIDUser
                .Add("@IDWorkGroupNew", SqlDbType.Int)
                .Item("@IDWorkGroupNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDWorkGroupNew").Value)
        End Using
    End Function

    Public Function WorkGroups_IsInUse(ByVal lvi As WorkGroupInfo) As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_IsInUse", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDWorkGroup", SqlDbType.Int).Value = lvi.IDWorkGroup
                .Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary
                .Add("@IsInUse", SqlDbType.Bit)
                .Item("@IsInUse").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteNonQuery()
            Return CBool(locCmd.Parameters("@IsInUse").Value)
        End Using

    End Function

    Public Sub WorkGroups_Delete(ByVal lvi As WorkGroupInfo)
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection

            'Assignments aufheben
            Dim locCmd As New SqlCommand("WorkGroups_DeleteAssignment", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary
                .Add("@IDWorkGroup", SqlDbType.Int).Value = lvi.IDWorkGroup
            End With
            locCmd.ExecuteNonQuery()

            'Workgroup löschen
            locCmd = New SqlCommand("WorkGroups_Delete", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDWorkGroup", SqlDbType.Int).Value = lvi.IDWorkGroup
                .Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary
            End With
            locCmd.ExecuteNonQuery()
        End Using
    End Sub

    Public Function WorkGroups_Edit(ByVal wgi As WorkGroupInfo, ByVal LastEditedByIDUser As Integer) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_Edit", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDWorkGroup", SqlDbType.Int).Value = wgi.IDWorkGroup
                .Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = wgi.IDCostCenter
                .Add("@WorkGroupNumber", SqlDbType.Int).Value = wgi.WorkGroupNumber
                .Add("@WorkGroupName", SqlDbType.NVarChar, 100).Value = wgi.WorkGroupName
                .Add("@WorkGroupDescription", SqlDbType.NVarChar, 4000).Value = wgi.WorkGroupDescription.Value
                .Add("@IsActive", SqlDbType.Bit).Value = wgi.IsActive
                .Add("@IsPeaceWork", SqlDbType.Bit).Value = wgi.IsPeaceWork
                .Add("@OrdinalNo", SqlDbType.Int).Value = wgi.OrdinalNo
                .Add("@IsConceptional", SqlDbType.Bit).Value = wgi.IsConceptional
                .Add("@TimeSettingDetails", SqlDbType.Xml).Value = wgi.TimeSettingDetails.XMLString
                .Add("@LastEditedByIDUser", SqlDbType.Int).Value = LastEditedByIDUser
                .Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = ConsiderHistoryMaintenance
                .Add("@IDWorkGroupNew", SqlDbType.Int)
                .Item("@IDWorkGroupNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDWorkGroupNew").Value)
        End Using
    End Function

    Friend Sub GetWorkGroupInfoCollection(ByVal CPInfo As CombinedParametersInfo, ByVal Wgi As WorkGroupInfoItems, ByVal WgiGetType As WorkGroupInfoItemsGetType)
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_GetItems", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDSubsidiary
                If CPInfo Is Nothing Then
                    .Add("@Shift", SqlDbType.Int).Value = DBNull.Value
                    .Add("@ProductionDate", SqlDbType.DateTime).Value = DBNull.Value
                Else
                    .Add("@Shift", SqlDbType.Int).Value = CPInfo.Shift
                    .Add("@ProductionDate", SqlDbType.DateTime).Value = CPInfo.ProductionDate
                End If
                If (WgiGetType And WorkGroupInfoItemsGetType.JoinedWithCostCenter) = WorkGroupInfoItemsGetType.JoinedWithCostCenter Then
                    .Add("@JoinedWithCostCenter", SqlDbType.Bit).Value = True
                Else
                    .Add("@JoinedWithCostCenter", SqlDbType.Bit).Value = False
                End If
            End With

            Dim locDr As SqlDataReader = locCmd.ExecuteReader()
            If locDr.HasRows Then
                Do While locDr.Read
                    Dim locWorkGroupInfo As New WorkGroupInfo(locDr, WgiGetType)
                    Wgi.Add(locWorkGroupInfo)
                Loop
            End If
        End Using
    End Sub

    Public Sub AssignLabourValuesToWorkGroup(ByVal IDSubsidiary As Integer, ByVal IDWorkGroup As Integer, ByVal LabourValues As LabourValueInfoCollection)
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        Dim locCount As Integer = 1
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_DeleteAssignment", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@IDWorkGroup", SqlDbType.Int).Value = IDWorkGroup
            End With
            locCmd.ExecuteNonQuery()

            For Each locLvi As LabourValueInfo In LabourValues
                locCmd = New SqlCommand("WorkGroups_AddAssignmentRecord", locConnection)
                locCmd.CommandType = CommandType.StoredProcedure

                With locCmd.Parameters
                    .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                    .Add("@IDLabourValueInternal", SqlDbType.Int).Value = locLvi.IDLabourValueInternal
                    .Add("@IDWorkGroup", SqlDbType.Int).Value = IDWorkGroup
                    .Add("@OrdinalNumber", SqlDbType.Int).Value = locCount
                End With
                locCmd.ExecuteNonQuery()
                locCount += 1
            Next
        End Using
    End Sub

    Friend Function WorkGroups_GetAssignedLabourValues(ByVal IDSubsidiary As Integer, ByVal IDWorkGroup As Integer) As LabourValueInfoCollection
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WorkGroups_GetAssignedLabourValues", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@IDWorkGroup", SqlDbType.Int).Value = IDWorkGroup
            End With
            Dim locDr As SqlDataReader = locCmd.ExecuteReader()
            If locDr.HasRows Then
                Dim locLic As New LabourValueInfoCollection
                Do While locDr.Read
                    Dim locLabourValueInfo As New LabourValueInfo(locDr, True)
                    locLic.Add(locLabourValueInfo)
                Loop
                Return locLic
            Else
                Return Nothing
            End If
        End Using
    End Function

    Friend Function GetWorkGroup(ByVal IDSubsidiary As Integer, ByVal IDWorkGroup As Integer) As WorkGroupInfo
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommand As New SqlCommand( _
                "SELECT [WorkGroups].*,CostCenters.CostCenterNo, CostCenters.CostCenterName, CostCenters.IncentiveIndicatorSynonym, CostCenters.IncentiveIndicatorDimension," & _
                "CostCenters.IncentiveIndicatorPrecision, CostCenters.IncentiveIndicatorFactor, CostCenters.BaseValuePrecision, CostCenters.BaseValueSynonym FROM [WorkGroups] " & _
                "[WorkGroups] INNER JOIN [CostCenters] ON " & _
                "[WorkGroups].[IDCostCenter] = [CostCenters].[IDCostCenter] WHERE " & _
                "[WorkGroups].[IDSubsidiary]=" & FacessoGeneric.LoginInfo.IDSubsidiary & _
                " AND [WorkGroups].[IsCurrent]='true' AND" & _
                "[WorkGroups].[IDWorkGroup]=" & IDWorkGroup, locConnection)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            If locDR.HasRows Then
                locDR.Read()
                Return New WorkGroupInfo(locDR, WorkGroupInfoItemsGetType.JoinedWithCostCenter)
            Else
                Return Nothing
            End If
        End Using
    End Function

    Friend Function GetWorkGroupByWorkGroupNumber(ByVal IDSubsidiary As Integer, ByVal WorkGroupNumber As Integer) As WorkGroupInfo
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommand As New SqlCommand( _
                "SELECT [WorkGroups].*,CostCenters.CostCenterNo, CostCenters.CostCenterName, CostCenters.IncentiveIndicatorSynonym, CostCenters.IncentiveIndicatorDimension," & _
                "CostCenters.IncentiveIndicatorPrecision, CostCenters.IncentiveIndicatorFactor, CostCenters.BaseValuePrecision, CostCenters.BaseValueSynonym FROM [WorkGroups] " & _
                "[WorkGroups] INNER JOIN [CostCenters] ON " & _
                "[WorkGroups].[IDCostCenter] = [CostCenters].[IDCostCenter] WHERE " & _
                "[WorkGroups].[IDSubsidiary]=" & FacessoGeneric.LoginInfo.IDSubsidiary & _
                " AND [WorkGroups].[IsCurrent]=1 AND" & _
                "[WorkGroups].[WorkGroupNumber]=" & WorkGroupNumber, locConnection)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            If locDR.HasRows Then
                locDR.Read()
                Return New WorkGroupInfo(locDR, WorkGroupInfoItemsGetType.JoinedWithCostCenter)
            Else
                Return Nothing
            End If
        End Using
    End Function



End Class

File: src/MixNetFx/Facesso.Data/SQLServer/SPAccess_WorkGroups.cs
using System;
using System.Data;
using System.Data.SqlClient;
using ActiveDev;
using Facesso;

namespace Facesso.Data
{
    public sealed partial class SPAccess
    {
        public bool WorkGroups_DoesWorkGroupNumberExist(int idSubsidiary, int workGroupNumber, ADDBNullable<int> excludeIDWorkGroup)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_DoesWorkGroupNumberExist", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@WorkGroupNumber", SqlDbType.Int).Value = workGroupNumber;
                locCmd.Parameters.Add("@ExcludeIDWorkGroup", SqlDbType.Int).Value = excludeIDWorkGroup.Value;
                locCmd.Parameters.Add("@DoesExist", SqlDbType.Bit);
                locCmd.Parameters["@DoesExist"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (bool)locCmd.Parameters["@DoesExist"].Value;
            }
        }

        public bool WorkGroups_DoesWorkGroupNameExist(int idSubsidiary, string workGroupName, ADDBNullable<int> excludeIDWorkGroup)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_DoesWorkGroupNameExist", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@WorkGroupName", SqlDbType.NVarChar, 20).Value = workGroupName;
                locCmd.Parameters.Add("@ExcludeIDWorkGroup", SqlDbType.Int).Value = excludeIDWorkGroup.Value;
                locCmd.Parameters.Add("@DoesExist", SqlDbType.Bit);
                locCmd.Parameters["@DoesExist"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (bool)locCmd.Parameters["@DoesExist"].Value;
            }
        }

        public int WorkGroups_Add(WorkGroupInfo wgi, int createdByIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_Add", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = wgi.IDCostCenter;
                locCmd.Parameters.Add("@WorkGroupNumber", SqlDbType.Int).Value = wgi.WorkGroupNumber;
                locCmd.Parameters.Add("@WorkGroupName", SqlDbType.NVarChar, 100).Value = wgi.WorkGroupName;
                locCmd.Parameters.Add("@WorkGroupDescription", SqlDbType.NVarChar, 4000).Value = wgi.WorkGroupDescription.Value;
                locCmd.Parameters.Add("@IsActive", SqlDbType.Bit).Value = wgi.IsActive;
                locCmd.Parameters.Add("@IsPeaceWork", SqlDbType.Bit).Value = wgi.IsPeaceWork;
                locCmd.Parameters.Add("@IsConceptional", SqlDbType.Bit).Value = wgi.IsConceptional;
                locCmd.Parameters.Add("@OrdinalNo", SqlDbType.Int).Value = wgi.OrdinalNo;
                locCmd.Parameters.Add("@TimeSettingDetails", SqlDbType.Xml).Value = wgi.TimeSettingDetails.XMLString();
                locCmd.Parameters.Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate;
                locCmd.Parameters.Add("@CreatedByIDUser", SqlDbType.Int).Value = createdByIDUser;
                locCmd.Parameters.Add("@IDWorkGroupNew", SqlDbType.Int);
                locCmd.Parameters["@IDWorkGroupNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDWorkGroupNew"].Value;
            }
        }

        public bool WorkGroups_IsInUse(WorkGroupInfo lvi)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_IsInUse", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = lvi.IDWorkGroup;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary;
                locCmd.Parameters.Add("@IsInUse", SqlDbType.Bit);
                locCmd.Parameters["@IsInUse"].Direction = ParameterDirection.Output;
                locCmd.ExecuteNonQuery();
                return (bool)locCmd.Parameters["@IsInUse"].Value;
            }
        }

        public void WorkGroups_Delete(WorkGroupInfo lvi)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                // Remove assignments first
                var locCmd = new SqlCommand("WorkGroups_DeleteAssignment", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = lvi.IDWorkGroup;
                locCmd.ExecuteNonQuery();

                // Delete workgroup
                locCmd = new SqlCommand("WorkGroups_Delete", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = lvi.IDWorkGroup;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary;
                locCmd.ExecuteNonQuery();
            }
        }

        public int WorkGroups_Edit(WorkGroupInfo wgi, int lastEditedByIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_Edit", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = wgi.IDWorkGroup;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = wgi.IDCostCenter;
                locCmd.Parameters.Add("@WorkGroupNumber", SqlDbType.Int).Value = wgi.WorkGroupNumber;
                locCmd.Parameters.Add("@WorkGroupName", SqlDbType.NVarChar, 100).Value = wgi.WorkGroupName;
                locCmd.Parameters.Add("@WorkGroupDescription", SqlDbType.NVarChar, 4000).Value = wgi.WorkGroupDescription.Value;
                locCmd.Parameters.Add("@IsActive", SqlDbType.Bit).Value = wgi.IsActive;
                locCmd.Parameters.Add("@IsPeaceWork", SqlDbType.Bit).Value = wgi.IsPeaceWork;
                locCmd.Parameters.Add("@OrdinalNo", SqlDbType.Int).Value = wgi.OrdinalNo;
                locCmd.Parameters.Add("@IsConceptional", SqlDbType.Bit).Value = wgi.IsConceptional;
                locCmd.Parameters.Add("@TimeSettingDetails", SqlDbType.Xml).Value = wgi.TimeSettingDetails.XMLString();
                locCmd.Parameters.Add("@LastEditedByIDUser", SqlDbType.Int).Value = lastEditedByIDUser;
                locCmd.Parameters.Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = FacessoGeneric.ConsiderHistoryMaintenance;
                locCmd.Parameters.Add("@IDWorkGroupNew", SqlDbType.Int);
                locCmd.Parameters["@IDWorkGroupNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDWorkGroupNew"].Value;
            }
        }

        internal void GetWorkGroupInfoCollection(CombinedParametersInfo cpInfo, WorkGroupInfoItems wgi, WorkGroupInfoItemsGetType wgiGetType)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_GetItems", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDSubsidiary;
                if (cpInfo == null)
                {
                    locCmd.Parameters.Add("@Shift", SqlDbType.Int).Value = DBNull.Value;
                    locCmd.Parameters.Add("@ProductionDate", SqlDbType.DateTime).Value = DBNull.Value;
                }
                else
                {
                    locCmd.Parameters.Add("@Shift", SqlDbType.Int).Value = cpInfo.Shift;
                    locCmd.Parameters.Add("@ProductionDate", SqlDbType.DateTime).Value = cpInfo.ProductionDate;
                }
                locCmd.Parameters.Add("@JoinedWithCostCenter", SqlDbType.Bit).Value =
                    (wgiGetType & WorkGroupInfoItemsGetType.JoinedWithCostCenter) == WorkGroupInfoItemsGetType.JoinedWithCostCenter;

                SqlDataReader locDr = locCmd.ExecuteReader();
                if (locDr.HasRows)
                {
                    while (locDr.Read())
                    {
                        var locWorkGroupInfo = new WorkGroupInfo(locDr, wgiGetType);
                        wgi.Add(locWorkGroupInfo);
                    }
                }
            }
        }

        public void AssignLabourValuesToWorkGroup(int idSubsidiary, int idWorkGroup, LabourValueInfoCollection labourValues)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            int locCount = 1;
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_DeleteAssignment", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = idWorkGroup;
                locCmd.ExecuteNonQuery();

                foreach (LabourValueInfo locLvi in labourValues)
                {
                    locCmd = new SqlCommand("WorkGroups_AddAssignmentRecord", locConnection);
                    locCmd.CommandType = CommandType.StoredProcedure;
                    locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                    locCmd.Parameters.Add("@IDLabourValueInternal", SqlDbType.Int).Value = locLvi.IDLabourValueInternal;
                    locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = idWorkGroup;
                    locCmd.Parameters.Add("@OrdinalNumber", SqlDbType.Int).Value = locCount;
                    locCmd.ExecuteNonQuery();
                    locCount++;
                }
            }
        }

        internal LabourValueInfoCollection WorkGroups_GetAssignedLabourValues(int idSubsidiary, int idWorkGroup)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WorkGroups_GetAssignedLabourValues", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = idWorkGroup;
                SqlDataReader locDr = locCmd.ExecuteReader();
                if (locDr.HasRows)
                {
                    var locLic = new LabourValueInfoCollection();
                    while (locDr.Read())
                    {
                        var locLabourValueInfo = new LabourValueInfo(locDr, true);
                        locLic.Add(locLabourValueInfo);
                    }
                    return locLic;
                }
                return null;
            }
        }

        internal WorkGroupInfo GetWorkGroup(int idSubsidiary, int idWorkGroup)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                var locCommand = new SqlCommand(
                    "SELECT [WorkGroups].*,CostCenters.CostCenterNo, CostCenters.CostCenterName, CostCenters.IncentiveIndicatorSynonym, CostCenters.IncentiveIndicatorDimension," +
                    "CostCenters.IncentiveIndicatorPrecision, CostCenters.IncentiveIndicatorFactor, CostCenters.BaseValuePrecision, CostCenters.BaseValueSynonym FROM [WorkGroups] " +
                    "[WorkGroups] INNER JOIN [CostCenters] ON " +
                    "[WorkGroups].[IDCostCenter] = [CostCenters].[IDCostCenter] WHERE " +
                    "[WorkGroups].[IDSubsidiary]=" + FacessoGeneric.LoginInfo.IDSubsidiary +
                    " AND [WorkGroups].[IsCurrent]='true' AND" +
                    "[WorkGroups].[IDWorkGroup]=" + idWorkGroup, locConnection);
                SqlDataReader locDR = locCommand.ExecuteReader();
                if (locDR.HasRows)
                {
                    locDR.Read();
                    return new WorkGroupInfo(locDR, WorkGroupInfoItemsGetType.JoinedWithCostCenter);
                }
                return null;
            }
        }

        internal WorkGroupInfo GetWorkGroupByWorkGroupNumber(int idSubsidiary, int workGroupNumber)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                var locCommand = new SqlCommand(
                    "SELECT [WorkGroups].*,CostCenters.CostCenterNo, CostCenters.CostCenterName, CostCenters.IncentiveIndicatorSynonym, CostCenters.IncentiveIndicatorDimension," +
                    "CostCenters.IncentiveIndicatorPrecision, CostCenters.IncentiveIndicatorFactor, CostCenters.BaseValuePrecision, CostCenters.BaseValueSynonym FROM [WorkGroups] " +
                    "[WorkGroups] INNER JOIN [CostCenters] ON " +
                    "[WorkGroups].[IDCostCenter] = [CostCenters].[IDCostCenter] WHERE " +
                    "[WorkGroups].[IDSubsidiary]=" + FacessoGeneric.LoginInfo.IDSubsidiary +
                    " AND [WorkGroups].[IsCurrent]=1 AND" +
                    "[WorkGroups].[WorkGroupNumber]=" + workGroupNumber, locConnection);
                SqlDataReader locDR = locCommand.ExecuteReader();
                if (locDR.HasRows)
                {
                    locDR.Read();
                    return new WorkGroupInfo(locDR, WorkGroupInfoItemsGetType.JoinedWithCostCenter);
                }
                return null;
            }
        }
    }
}
